pkg/auth/providers/saml/saml_provider.go

```diff
@@ -0,0 +1,287 @@
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/apis/management.cattle.io/v3public"
+	"github.com/rancher/types/client/management/v3"
+	publicclient "github.com/rancher/types/client/management/v3public"
+	"github.com/rancher/types/config"
+	"github.com/rancher/types/user"
+	"github.com/sirupsen/logrus"
+	log "github.com/sirupsen/logrus"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	Name      = "saml"
+	userType  = Name + "_user"
+	groupType = Name + "_group"
+)
+
+type Provider struct {
+	ctx         context.Context
+	authConfigs v3.AuthConfigInterface
+	userMGR     user.Manager
+	samlClient  *SPClient
+}
+
+var Samlp *Provider
+
+func (s *Provider) GetName() string {
+	return Name
+}
+
+func Configure(ctx context.Context, mgmtCtx *config.ScaledContext, userMGR user.Manager) common.AuthProvider {
+	samlClient := &SPClient{}
+	Samlp = &Provider{
+		ctx:         ctx,
+		authConfigs: mgmtCtx.Management.AuthConfigs(""),
+		userMGR:     userMGR,
+		samlClient:  samlClient,
+	}
+	return Samlp
+}
+
+func (s *Provider) CustomizeSchema(schema *types.Schema) {
+	schema.ActionHandler = s.actionHandler
+	schema.Formatter = s.formatter
+}
+
+func (s *Provider) TransformToAuthProvider(authConfig map[string]interface{}) map[string]interface{} {
+	p := common.TransformToAuthProvider(authConfig)
+	p[publicclient.SamlProviderFieldRedirectURL] = formSamlRedirectURLFromMap(authConfig)
+	return p
+}
+
+func (s *Provider) AuthenticateUser(input interface{}) (v3.Principal, []v3.Principal, map[string]string, error) {
+	login, ok := input.(*v3public.CodeBasedLogin)
+	if !ok {
+		return v3.Principal{}, nil, nil, errors.New("SAML: unexpected input type")
+	}
+
+	return s.LoginUser(login, nil, false)
+}
+
+func (s *Provider) LoginUser(samlCredential *v3public.CodeBasedLogin, config *v3.SamlConfig, test bool) (v3.Principal, []v3.Principal, map[string]string, error) {
+	var groupPrincipals []v3.Principal
+	var userPrincipal v3.Principal
+	//var providerInfo = make(map[string]string)
+	var err error
+	var samlData map[string][]string
+
+	if config == nil {
+		config, err = s.GetSamlConfig()
+		if err != nil {
+			return v3.Principal{}, nil, nil, err
+		}
+	}
+	s.samlClient.config = config
+	inputMap := samlCredential.Code
+
+	if err := json.Unmarshal([]byte(inputMap), &samlData); err != nil {
+		log.Errorf("SAML: Error getting saml data from input %v", err)
+		return v3.Principal{}, nil, nil, err
+	}
+
+	accounts, err := s.samlClient.getShibIdentities(samlData)
+	if err != nil {
+		return v3.Principal{}, nil, nil, err
+	}
+
+	for _, a := range accounts {
+		if !a.IsGroup {
+			userPrincipal = s.toPrincipal(userType, a, nil)
+			userPrincipal.Me = true
+		} else {
+			groupPrincipal := s.toPrincipal(groupType, a, nil)
+			groupPrincipal.MemberOf = true
+			groupPrincipals = append(groupPrincipals, groupPrincipal)
+		}
+	}
+
+	testAllowedPrincipals := config.AllowedPrincipalIDs
```

Comment:
I know this was just copied from another provider, but the name `testAllowedPrincipals` is confusing. The word "test" doesnt make sense to me. Can you change it to just `allowedPrincipals`